pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/generator/BaseTaskGenerator.java

```diff
@@ -117,6 +118,26 @@ public List<SegmentZKMetadata> getSegmentsZKMetadataForTable(String tableNameWit
     return selectedSegmentZKMetadataList;
   }
 
+  public List<SegmentZKMetadata> getNonConsumingSegmentsZKMetadataForRealtimeTable(String tableNameWithType) {
+    IdealState idealState = _clusterInfoAccessor.getIdealState(tableNameWithType);
+    Set<String> idealStateSegments = idealState.getPartitionSet();
+    List<SegmentZKMetadata> segmentZKMetadataList = _clusterInfoAccessor.getSegmentsZKMetadata(tableNameWithType);
+    List<SegmentZKMetadata> selectedSegmentZKMetadataList = new ArrayList<>();
+    for (SegmentZKMetadata segmentZKMetadata : segmentZKMetadataList) {
+      String segmentName = segmentZKMetadata.getSegmentName();
+      if (idealStateSegments.contains(segmentName)
+          && segmentZKMetadata.getStatus().isCompleted() // skip consuming segments
+          && !idealState.getInstanceStateMap(segmentName).containsValue(SegmentStateModel.CONSUMING)) {
```

Comment:
Can this be done by going through the idealStates for each online segment and just fetch the zkmetadata from getSegmentsZKMetadata(tableNameWithType, segmentName)? In that case we would not need the 3rd check 
`idealState.getInstanceStateMap(segmentName).containsValue(SegmentStateModel.CONSUMING)`

Reply:
To find the list of online segments, we need to use `.containsValue(ONLINE)`, which makes it again 3 checks!